### ISSUE TITLE
The calculation of MSE is not correct

### ISSUE THREAD

**Issue by Jillian2017** [tensorflow/compression#12]
Hi, Johannes:
      Thanks a lot for your released code. There is a small error in your calculation of MSE, since the num_pixels = batchsize* patchsize*patchsize, which is reasonable for bpp, but for MSE, it should be num_pixels = batchsize* patchsize*patchsize*3, or you can use tf.reduce_mean instead. 
     What's more,  I find that the R-D loss curve oscillates a lot in training phase, and it does not converge with small lambda(such as 0.001). Can you give me some advice to train the model well.
     Best regards.

**Comment by jonycgn** [comment]
Hi Jillian,

I'm not sure why you think that mean squared error must be averaged across the RGB channels. The current code treats the RGB information as a colorspace. The computed mean squared error is then the squared Euclidean distance in this 3-dimensional colorspace, averaged across pixels. I can't see anything fundamentally wrong with that, it is simply a different convention (which makes sense for other colorspaces as well).

When you set lambda to a very small number, eventually, the bit rate should end up collapsing to 0 bpp, and the distortion will be very large. This should be expected. Have you tried raising lambda? What is the smallest bpp that you can successfully train?

**Comment by Jillian2017** [comment]
Hi, Johannes:
     Thanks for your explanation.
     Yes, I agree with that this is not fundamentally error, and I think this make little difference in training. In my opinion, the mean squared error should be averaged across points instead of RGB channels. Suppose input and output are not images with 3 channels, they are arrays that are made up of some elements, then the MSE should be the average value across all these points.
    For the problem of convergence, I have tried to train the model with lambda equals 0.001, the trend of the RD curve is downward actually, after certain iteration(maybe a hundred thousand) the MSE value oscillates between 150 and 600,   and it will not converge even the last_step is very high(1e7). My training data contains about 2 thousand patches, the learning rate is the same. Is data augmentation needed for small bitrate? As for higher lambda, such as 0.1, it does not oscillate a lot.
     Best regards.

**Comment by Jillian2017** [comment]
Hi, Johannes:
     Thanks a lot. Very appreciate your reply and look forward to your report.
     Best regards.

**Comment by jonycgn** [comment]
Jillian,

I was able to train models at quite low bitrates just fine (lambda=0.001 and lambda=0.00025). I don't see any oscillation happening, or any other problems with convergence.

From what dataset did you gather your image patches for training?

Could you check out the latest code, try again, and see if the problem is still happening?

**Comment by Jillian2017** [comment]
Johannes:
      Thanks a lot for your timely reply. I have changed the way of reading images, in my side, 256*256 patches are randomly cropped from high resolution images. What's more, my dataset contains less than 3 thousand high resolution images which are made up of DIV2K, CLIC's training images. So, is it possible that there are not  much images in dataset. There are many high resolution in website, but most of them are in lossy format, such as jpeg, I think they are not good candidates for image compression. Do you have any suggestion for data augment based on these lossy images?
   What's more I will download your lastest code  and train again.
   Best regards.

**Comment by Jillian2017** [comment]
Hi, Johannes:
     I haved downloaded your latest code and run it again, but the result seems not improve. My training dataset contains almost 3000 thousands images in png format, lambda is 0.001, patch size is 256*256, and number filters is 192. After 526144 steps, the loss still oscillates, attached is the screenshot from tensorboard.
![screenshot at 2019-01-01 19-30-52](https://user-images.githubusercontent.com/30519938/50572576-4b700e80-0dfe-11e9-8184-930b374d0b31.png)
    Wish your advice. Best regards.

**Comment by jonycgn** [comment]
Hi, these curves look completely normal. What you are seeing is the natural fluctuation of the loss due to stochastic gradients. Try running the model on one of the Kodak images (with "python bls2017.py -v compress") and compare the results to the corresponding plot in the appendix of [this paper](https://arxiv.org/abs/1802.01436). You should get something matching the Balle 2017 curve quite closely.

I'm closing the bug for now. If your results are way off, and you still think there is a bug in the library, you can reopen it. If you need tips for tweaking the dataset/training, please post to the Google group.

Again, thanks for pointing out the MSE inconsistency!
